agent/xds/routes.go

```diff
@@ -430,6 +425,88 @@ func (s *ResourceGenerator) routesForIngressGateway(cfgSnap *proxycfg.ConfigSnap
 	return result, nil
 }
 
+// routesForAPIGateway returns the xDS API representation of the
+// "routes" in the snapshot.
+func (s *ResourceGenerator) routesForAPIGateway(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
+	var result []proto.Message
+
+	readyUpstreamsList := getReadyUpstreams(cfgSnap)
+
+	for _, readyUpstreams := range readyUpstreamsList {
+		listenerCfg := readyUpstreams.listenerCfg
+		// Do not create any route configuration for TCP listeners
+		if listenerCfg.Protocol == "tcp" {
+			continue
+		}
+
+		routeRef := readyUpstreams.routeReference
+		listenerKey := readyUpstreams.listenerKey
+
+		// Depending on their TLS config, upstreams are either attached to the
+		// default route or have their own routes. We'll add any upstreams that
+		// don't have custom filter chains and routes to this.
+		defaultRoute := &envoy_route_v3.RouteConfiguration{
+			Name: listenerKey.RouteName(),
+			// ValidateClusters defaults to true when defined statically and false
+			// when done via RDS. Re-set the reasonable value of true to prevent
+			// null-routing traffic.
+			ValidateClusters: makeBoolValue(true),
+		}
+
+		route, ok := cfgSnap.APIGateway.HTTPRoutes.Get(routeRef)
+		if !ok {
+			return nil, fmt.Errorf("missing route for route reference %s:%s", routeRef.Name, routeRef.Kind)
+		}
+
+		flattenedRoutes := discoverychain.FlattenHTTPRoute(route, &listenerCfg, cfgSnap.APIGateway.GatewayConfig)
+
+		for _, flattenedRoute := range flattenedRoutes {
+			flattenedRoute := flattenedRoute
+
+			upstream := buildHTTPRouteUpstream(flattenedRoute, listenerCfg)
+			uid := proxycfg.NewUpstreamID(&upstream)
+			chain := cfgSnap.APIGateway.DiscoveryChain[uid]
+			if chain == nil {
+				s.Logger.Debug("Discovery chain not found for flattened route", "discovery chain ID", uid)
+				continue
+			}
+
+			domains := generateUpstreamAPIsDomains(listenerKey, upstream, flattenedRoute.Hostnames)
+
+			virtualHost, err := s.makeUpstreamRouteForDiscoveryChain(cfgSnap, uid, chain, domains, false)
+			if err != nil {
+				return nil, err
+			}
+
+			injectHeaderManipToVirtualHostAPIGateway(&flattenedRoute, virtualHost)
+
+			// TODO Handle TLS config and add new route if appropriate
+			//   We need something analogous to routeNameForUpstream used below
+			//   But currently ToIngress is not handeling this usecase
+			defaultRoute.VirtualHosts = append(defaultRoute.VirtualHosts, virtualHost)
```

Comment:
Does this mean we're not generating TLS configuration for listeners where it has been configured and expected? Would be good to double-check test coverage for this to confirm that we're okay with the current behavior...

Reply:
It does currently generate TLS configuration, but there is an option to handle SDSOverrides in ingress gateway that we don't support currently that I believe this TODO was referring to

Reply:
We just need to confirm that we don't use the additional fields that `routeNameForUpstream` operates on anywhere, ever then we can remove this TODO

Reply:
We confirmed that this TODO can be deleted. It always led to the logic we have here because we were never setting SDS overrides when using `ToIngress()`. All good!